### PR TITLE
Clear stale chosen_shipping_methods session when frontend sends empty shipping

### LINKED ISSUE #51197: Session variable 'chosen_shipping_methods' not emptied when no more shipping method chosen
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

- A cart contains both virtual and non virtual products
- A shipping method is chosen on the checkout page, 
- You update your cart, and you are in the case where no shipping method is needed anymore because all your products in the cart are now virtual
- Then you can see in your code that session variable 'chosen_shipping_methods' is not emptied

### Expected behavior

If there are no more shipping method needed for the checkout when updating the cart, the session variable should be emptied.

### Actual behavior

In https://github.com/woocommerce/woocommerce/blob/c884a1e70e92faf1253dc765534122912f4a0102/plugins/woocommerce/includes/class-wc-ajax.php#L354, we are in the Class WC_AJAX, function update_order_review()

we have the lines of code that get $posted_shipping_methods and then update $chosen_shipping_methods, then set the session variable : 
`WC()->session->set( 'chosen_shipping_methods', $chosen_shipping_methods );`

If $posted_shipping_methods  is an empty array, we should set $chosen_shipping_methods to empty as well.
But we only have : 

`
if ( is_array( $posted_shipping_methods ) ) {
			foreach ( $posted_shipping_methods as $i => $value ) {
				if ( ! is_string( $value ) ) {
					continue;
				}
				$chosen_shipping_methods[ $i ] = $value;
			}
		}`

which updates the array, but can't empty it.

### Steps to reproduce

Steps are above.

Extract of the code used to see the session variable

`add_filter( 'woocommerce_available_payment_gateways', 'dw_turn_off_cheque_virt_devis' );  
function dw_turn_off_cheque_virt_devis( $available_gateways ) {	
	if ((is_checkout()  && !is_wc_endpoint_url('order-pay') && WC()->session)) { // get shipping methode used
		$chosen_methods = WC()->session->get('chosen_shipping_methods');
error_log(var_export($chosen_methods, true));
       }
       return $available_gateways 
}`

I corrected the bug in my theme by adding this code

`add_action( 'woocommerce_checkout_update_order_review', 'dw_remove_session_empty_shipping_data', 20);
function  dw_remove_session_empty_shipping_data($post_data) {	
	$posted_shipping_methods = isset( $_POST['shipping_method'] ) ? wc_clean( wp_unslash( $_POST['shipping_method'] ) ) : array();
	if (  is_array( $posted_shipping_methods ) && count($posted_shipping_methods) == 0) {
		WC()->session->set( 'chosen_shipping_methods', array() );
		WC()->session->set( 'chosen_payment_method', '');
	}	
}`

### WordPress Environment

`
### WordPress Environment ###

WC Version: 8.8.5
REST API Version: ✔ 8.8.5
Action Scheduler Version: ✔ 3.7.4
Log Directory Writable: ✔
WP Version: 6.6.1
WP Multisite: –
WP Memory Limit: 256 Mo
WP Debug Mode: ✔
WP Cron: ✔
Language: fr_FR
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.57 (Debian)
PHP Version: 8.2.15
PHP Post Max Size: 64 Mo
PHP Time Limit: 30
PHP Max Input Vars: 1000
cURL Version: 7.88.1
OpenSSL/3.0.11

SUHOSIN Installed: –
MySQL Version: 10.4.33-MariaDB-1:10.4.33+maria~ubu2004
Max Upload Size: 64 Mo
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 8.8.5
WC Database Prefix: wp_
Taille totale de la base de données: 91.73MB
Taille de la base de données: 63.91MB
Taille de l’index: 27.82MB

### Post Type Counts ###

page: 21
post: 152
product: 651
product_variation: 924
revision: 1662
shop_coupon: 4
shop_order: 103

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (18) ###

Gravity Forms: par Gravity Forms – 2.8.12
Advanced Custom Fields PRO: par WP Engine – 6.2.9
WooCommerce Demande de devis: par PerfectoTech – 1.0.0
EWWW Image Optimizer: par Exactement WWW – 7.6.0
Fake Pay For WooCommerce: par Anthony Graddy – 1.1.0
GDPR Cookie Compliance: par Moove Agency – 4.14.0
Gravity Forms reCAPTCHA Add-On: par Gravity Forms – 1.5.0
JSM Non-Breaking Space (nbsp) for French Content: par JS Morisset – 1.13.1
SMTP Mailer: par naa986 – 1.1.13
UpdraftPlus - Sauvegarde/Restauration: par UpdraftPlus.Com
DavidAnderson – 1.24.3

User Switching: par John Blackbourn – 1.8.0
Systempay for WooCommerce: par Lyra Network – 1.12.2
WooCommerce Extended Coupon Features FREE: par Soft79 (Soft79) – 3.3.3
WooCommerce Table Rate Shipping: par Bolder Elements – 4.2.1
WooCommerce: par Automattic – 8.8.5 (une mise à jour de la version 9.2.3 est disponible)
Yoast SEO: par L’équipe Yoast – 22.7
WP Crontrol: par John Blackbourn – 1.16.3
WP Help: par Mark Jaquith – 1.7.5

### Settings ###

API Enabled: –
Force SSL: –
Currency: EUR (€)
Currency Position: right_space
Thousand Separator: 
Decimal Separator: ,
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔
HPOS feature screen enabled: –
HPOS feature enabled: –
Order datastore: WC_Order_Data_Store_CPT
HPOS data sync enabled: –

### Logging ###

Enabled: ✔
Handler: Automattic\WooCommerce\Internal\Admin\Logging\LogHandlerFileV2
Retention period: 30 jours
Level threshold: –
Log directory size: 242 Ko

### WC Pages ###

Base de la boutique: #7 - /librairie/
Panier: #8 - /panier/
Commander: #9 - /commander/
Mon compte: #10 - /mon-compte/
Conditions générales de vente et d’utilisation: #11 - /remboursements_retours/

### Theme ###

Child Theme: ✔
Parent Theme Name: Storefront
Parent Theme Version: 4.4.1 – 4.6.0 est disponible
Parent Theme Author URL: https://woocommerce.com/
WooCommerce Support: ✔

### Templates ###

Overrides: /var/www/html/wp-content/plugins/woocommerce/templates/block-notices/error.php
/var/www/html/wp-content/plugins/woocommerce/templates/block-notices/notice.php
/var/www/html/wp-content/plugins/woocommerce/templates/block-notices/success.php
docubook/parts/checkout-header.html
docubook/parts/mini-cart.html
docubook/templates/archive-product.html
docubook/templates/order-confirmation.html
docubook/templates/product-search-results.html
docubook/templates/single-product.html
docubook/templates/taxonomy-product_attribute.html
docubook/templates/taxonomy-product_cat.html
docubook/templates/taxonomy-product_tag.html

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customize-store
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
product-variation-management
product-virtual-downloadable
product-external-affiliate
product-grouped
product-linked
product-pre-publish-modal
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: minified-js
new-product-management-experience
product-custom-fields
settings
async-product-editor-category-field
launch-your-store

Daily Cron: ✔ Next scheduled: 2024-09-07 08:22:17 +02:00
Options: ✔
Notes: 63
Onboarding: completed

### Action Scheduler ###

Terminé: 12 762
Oldest: 2024-08-06 15:20:24 +0200
Newest: 2024-09-06 14:31:27 +0200

Échoué: 23
Oldest: 2023-09-06 12:05:59 +0200
Newest: 2024-05-25 14:38:51 +0200

En attente: 2
Oldest: 2024-09-06 15:38:16 +0200
Newest: 2024-09-06 16:41:06 +0200

### Status report information ###

Generated at: 2024-09-06 15:05:47 +02:00
`

### Isolating the problem

- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by shsajalchowdhury**
## Submission Review Guidelines
- [x] I have read the Contributing Guidelines for this project.
- [x] I agree to follow the Code of Conduct that this project has adopted.

## Changes proposed in this Pull Request
Closes #51197.

When a cart contains both virtual and physical products and the user has selected a shipping method, removing all physical products leaves the `chosen_shipping_methods` session variable holding stale shipping method values. This is because the frontend sends an empty `shipping_method` POST array, but the existing code only iterates over it (doing nothing for an empty array) without clearing the previous session values.

This PR resets `$chosen_shipping_methods` to an empty array before the foreach loop when the `shipping_method` key is present in the POST data. The loop then repopulates it with any valid values. If the frontend sends an empty array (all products are now virtual), the session is correctly cleared.

Both occurrences in `class-wc-ajax.php` are fixed:
- `update_order_review()` (cart update)
- `checkout()` (checkout update)

## How to test the changes in this Pull Request
1. Create a virtual product and a physical product
2. Add both to cart
3. Go to checkout, select a shipping method
4. Remove the physical product from cart
5. Observe: `WC()->session->get('chosen_shipping_methods')` should now be empty
6. **Before fix:** Session still contains the old shipping method
7. **After fix:** Session is correctly cleared

## Testing that has already taken place
- PHP syntax check passes
- phpcs lint: no new errors (pre-existing issues only)
- Logic: reset before rebuild pattern ensures no stale data

## Changelog
- [x] Automatically create on merge
- Significance: Patch
- Type: Fix
- Message: Clear stale chosen_shipping_methods session when no shipping methods are needed

### CODE REVIEW

File: plugins/woocommerce/includes/class-wc-ajax.php [hunk 1]
```diff
@@ -406,6 +410,10 @@ public static function update_order_review() {
 		$chosen_shipping_methods = WC()->session->get( 'chosen_shipping_methods' );
 		$posted_shipping_methods = isset( $_POST['shipping_method'] ) ? wc_clean( wp_unslash( $_POST['shipping_method'] ) ) : array();
 
+		if ( ! isset( $_POST['shipping_method'] ) ) {
```
Comment (mikejolley) [inline plugins/woocommerce/includes/class-wc-ajax.php:413]: This treats every request without a top-level `shipping_method` key as an empty shipping update. In checkout, that key is also omitted intentionally when `update_checkout` runs with `update_shipping_method: false` (e.g. coupon apply/remove), so those refreshes would clear the existing `chosen_shipping_methods` session even though shipping wasn't changed.

**Review by mikejolley — CHANGES_REQUESTED** [review]
Thanks for working on this. I agree the stale `chosen_shipping_methods` session needs fixing, but I don't think we can merge this version.

The main issue is in `update_order_review()`: an omitted `shipping_method` currently means two different things. It can mean “there are no shipping methods to send”, but `checkout.js` also omits it intentionally when `update_checkout` runs with `update_shipping_method: false` (for example after applying or removing a coupon). With this change, those checkout refreshes would clear the existing `chosen_shipping_methods` session even though the customer didn't change shipping.

There's also no regression coverage for that empty-versus-omitted distinction. We'd need tests proving that an empty shipping update clears stale methods, and that an omitted `shipping_method` caused by `update_shipping_method: false` preserves the current choice.

I'm going to close this PR for now because the current approach introduces that regression risk. We're happy to look again if it's reopened with a fix that distinguishes those cases and includes the coverage above.
